File: 820-find-eventual-safe-states/find-eventual-safe-states.cpp
class Solution {
public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        unordered_map<int, bool> safe;
        vector<int> res;

        function<bool(int)> dfs = [&](int i) {
            if (safe.find(i) != safe.end()) {
                return safe[i];
            }
            safe[i] = false;
            for (int nei : graph[i]) {
                if (!dfs(nei)) {
                    return false;
                }
            }
            safe[i] = true;
            return true;
        };

        for (int i = 0; i < n; ++i) {
            if (dfs(i)) {
                res.push_back(i);
            }
        }
        return res;
    }
};

File: 820-find-eventual-safe-states/find-eventual-safe-states.java
class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        int n = graph.length;
        Map<Integer, Boolean> safe = new HashMap<>();
        List<Integer> res = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            if (dfs(i, graph, safe)) {
                res.add(i);
            }
        }
        return res;
    }

    private boolean dfs(int i, int[][] graph, Map<Integer, Boolean> safe) {
        if (safe.containsKey(i)) {
            return safe.get(i);
        }
        safe.put(i, false);
        for (int nei : graph[i]) {
            if (!dfs(nei, graph, safe)) {
                return false;
            }
        }
        safe.put(i, true);
        return true;
    }
}